js/exam.js

```diff
@@ -0,0 +1,714 @@
+
+  let response = await AJAXCall({
+    phpFilePath: "../include/exam/addNewExam.php",
+    rejectMessage: "New Exam Failed To Add",
+    params,
+    type: "post",
+  });
+
+  console.log("exam generation response: ", response);
+
+  fetchAllExam(courseID);
+
+  setTimeout(() => {
+    closeExamModal();
+    removeLoader(loader);
+  }, 2000);
+}
+
+async function saveExamAsJSON(filename, ArrayContainingObjects, type) {
+  let JSONString = JSON.stringify(ArrayContainingObjects);
+
+  let correctPath;
+
+  switch (type) {
+    case "student":
+    case "new":
+    case "resume":
+      correctPath = `../exam/taken/${filename}`;
+      break;
+    case "teacher":
+    case "generated":
+      correctPath = `../exam/generated/${filename}`;
+      break;
+  }
+
+  console.log("[3] correctPath: ", correctPath);
+  console.log("[4] jsonString: ", JSONString);
+
+  try {
+    let result = await AJAXCall({
+      phpFilePath: "../include/saveJSONData.php",
+      rejectMessage: "saving json file failed",
+      params: `filepath=${correctPath}&&jsonString=${JSONString}`,
+      type: "post",
+    });
+
+    console.log("[5] async Result: ", result);
+  } catch (error) {
+    //TODO: bubbleUpError()
+    console.log(error);
+  }
+}
+
+async function fetchAllExam(id) {
+  const examResponse = await AJAXCall({
+    phpFilePath: "../include/exam/getAllExam.php",
+    rejectMessage: "Get All Exam Failed To Be Fetched",
+    params: `courseID=${id}`,
+    type: "fetch",
+  });
+
+  const examsContainer = document.getElementById("exams-container");
+  examsContainer.innerHTML = "";
+
+  for (let i = 0; i < examResponse.length; i++) {
+    const newDiv = document.createElement("div");
+    const newFirstP = document.createElement("p");
+    const newSecondP = document.createElement("p");
+
+    newDiv.className = "exam-item";
+
+    newFirstP.innerHTML = examResponse[i].examName;
+    newSecondP.innerHTML = examResponse[i].examDate;
+
+    newDiv.appendChild(newFirstP);
+    newDiv.appendChild(newSecondP);
+    examsContainer.appendChild(newDiv);
+  }
+
+  if (examResponse.length === 0) {
+    const newDiv = document.createElement("div");
+    newDiv.className = "exam-item";
+    newDiv.style = "display:flex; justify-content:center;align-items:center;";
+
+    const newFirstP = document.createElement("p");
+
+    newFirstP.innerHTML = "You did not create any exam!";
+    newDiv.appendChild(newFirstP);
+    examsContainer.appendChild(newDiv);
+  }
+}
+
+async function getAllCoursesOfStudent() {
+  let { id: globalUserID } = await globalUserDetails;
+
+  const subscriptions = await AJAXCall({
+    phpFilePath: "../include/exam/getAllSubscriptions.php",
+    rejectMessage: "Get All Subscriptions Failed To Be Fetched",
+    params: `userID=${globalUserID}`,
+    type: "fetch",
+  });
+
+  const exams = [];
+
+  for (let i = 0; i < subscriptions.length; i++) {
```

Comment:
For loops like this can cause major errors because you aren't catching any errors for subsequent ajax calls. Assuming this loop runs 10 times. What if on the 7th loop run, the AJAX call fails. This is what is happening to me right now. The loop call failed and it's not showing anything.